lib/go/test/flow_epoch_test.go

```diff
@@ -1516,3 +1517,555 @@ func TestEpochReset(t *testing.T) {
+func TestEpochRecover(t *testing.T) {
+
+	// Perform epoch recovery with a new epoch and epoch recover overwriting the current epoch.
+	t.Run("Can recover the epoch and have everything return to normal", func(t *testing.T) {
+		epochConfig := &testEpochConfig{
+			startEpochCounter:    startEpochCounter,
+			numEpochViews:        numEpochViews,
+			numStakingViews:      numStakingViews,
+			numDKGViews:          numDKGViews,
+			numClusters:          numClusters,
+			numEpochAccounts:     numEpochAccounts,
+			randomSource:         randomSource,
+			rewardIncreaseFactor: rewardIncreaseFactor,
+		}
+		runWithDefaultContracts(t, epochConfig, func(b emulator.Emulator, env templates.Environment, ids []string, idTableAddress flow.Address, IDTableSigner sdkcrypto.Signer, adapter *adapters.SDKAdapter) {
+			advanceView(t, b, env, idTableAddress, IDTableSigner, 1, "EPOCHSETUP", false)
+			epochTimingConfigResult := executeScriptAndCheck(t, b, templates.GenerateGetEpochTimingConfigScript(env), nil)
+			var (
+				startView      uint64 = 100
+				stakingEndView uint64 = 120
+				endView        uint64 = 160
+				targetDuration uint64 = numEpochViews
+				epochCounter   uint64 = startEpochCounter + 1
+				targetEndTime  uint64 = expectedTargetEndTime(epochTimingConfigResult, epochCounter)
+			)
+			args := getRecoveryTxArgs(env, ids, startView, stakingEndView, endView, targetDuration, targetEndTime, epochCounter)
+
+			tx := createTxWithTemplateAndAuthorizer(b, templates.GenerateRecoverEpochScript(env), idTableAddress)
+			for _, arg := range args {
+				tx.AddArgument(arg)
+			}
+
+			signAndSubmit(
+				t, b, tx,
+				[]flow.Address{idTableAddress},
+				[]sdkcrypto.Signer{IDTableSigner},
+				false,
+			)
+
+			advanceView(t, b, env, idTableAddress, IDTableSigner, 1, "BLOCK", false)
+
+			verifyEpochRecoverGovernanceTx(t, b, env, ids,
+				startView,
+				stakingEndView,
+				endView,
+				targetDuration,
+				targetEndTime,
+				epochCounter,
+				"0.0",
+				idTableAddress,
+				adapter,
+				args,
+			)
+		})
+	})
+
+	t.Run("Can recover the current epoch and overwrite the current epoch metadata", func(t *testing.T) {
+		epochConfig := &testEpochConfig{
+			startEpochCounter:    startEpochCounter,
+			numEpochViews:        numEpochViews,
+			numStakingViews:      numStakingViews,
+			numDKGViews:          numDKGViews,
+			numClusters:          numClusters,
+			numEpochAccounts:     numEpochAccounts,
+			randomSource:         randomSource,
+			rewardIncreaseFactor: rewardIncreaseFactor,
+		}
+		runWithDefaultContracts(t, epochConfig, func(b emulator.Emulator, env templates.Environment, ids []string, idTableAddress flow.Address, IDTableSigner sdkcrypto.Signer, adapter *adapters.SDKAdapter) {
+			// Advance to epoch Setup and make sure that the epoch cannot be ended
+			advanceView(t, b, env, idTableAddress, IDTableSigner, 1, "EPOCHSETUP", false)
```

Comment:
Unless I'm missing one, all of the test cases transition into the EpochSetup phase prior to executing the recovery. I would like to have at least one test case where we execute the recovery while in the Staking phase.

Reply:
added test cases for both phases in https://github.com/onflow/flow-core-contracts/pull/440/commits/629faa7fa778f1f5b9f61de0fa9100f94e6f1b52